Implement NoImplicitAnyManipulator

guor8lei opened the pull request:
## Issue
Fixes #27 

## Description
Implements `NoImplicitAnyManipulator` to fix for variable and parameter cases of `--noImplicitAny`. Namely, implements the basic topological sorting fix as described in my deep dive presentation. 
1. First constructs a set of declarations that are implicitly any.
2. Constructs a dependency graph based on the values and expressions each variable is assigned to. If the variable is a parameter, also looks for all function calls and extracts the function arguments' types and expressions.
3. Topologically sorts the graph.
4. In topological order, assigns the types to each declaration. If a declaration has type any, console logs and skips the declaration, as well all declarations that rely on it.

I moved a few functions from `StrictNullChecksManipulator` to `Manipulator` as their functionality was shared between `StrictNullChecksManipulator` and `NoImplicitAnyManipulator`.

Note: This implementation does _not_ fix for more complex cases involving empty arrays or objects, as those cases must be fixed with hardcoded implementations, which may be dangerous if not every cases is taken into account.

## Tests
Unit Tests:
- Topologically sorts graph
- Constructs descendants map
- Calculates declaration types given declaration dependencies
- Adds assignment dependencies
- Adds function call dependencies

Functional Tests:
- Assigns types based on assignments to literals / expressions
- Assigns types based on function calls
- Assigns types based on dependencies between declarations (topologically sorting)
- Prevents `any` pollution

jiady commented on src/manipulators/no_implicit_any_manipulator.ts:
```diff
@@ -26,9 +37,286 @@ export class NoImplicitAnyManipulator extends Manipulator {
+      references?.forEach(reference => {
+        const parent = reference.getParentIfKind(SyntaxKind.BinaryExpression);
+        const sibling = reference.getNextSiblingIfKind(SyntaxKind.EqualsToken);
+        const nextSibling = sibling?.getNextSibling();
```
let's use more meaningful variable name.

parent/sibling/nextSibling may be meaningful from AST perspective, can we use more straight forward language like 
parent -> binaryExpressionNode
sibling-> equlaTokenNode
nextSibling-> assignmentValueNode

etc

so is others.

guor8lei replied:
renamed variables

michaelyara commented on src/manipulators/no_implicit_any_manipulator.ts:
```diff
@@ -26,9 +37,286 @@ export class NoImplicitAnyManipulator extends Manipulator {
+      numDependencies.set(vertex, 0);
+    }
+
+    for (const successors of edges.values()) {
```
suggestion: you may be able to use flatMap here

guor8lei replied:
I ended up reimplementing `topoSort` to handle cyclical dependencies better

michaelyara commented on src/manipulators/no_implicit_any_manipulator.ts:
```diff
@@ -26,9 +37,286 @@ export class NoImplicitAnyManipulator extends Manipulator {
+    // Add all vertices with in-degree to queue.
+    const queue: T[] = [];
+    for (const vertex of numDependencies.keys()) {
+      if (numDependencies.get(vertex) === 0) {
```
suggestion: can be a filter

guor8lei replied:
I ended up reimplementing topoSort to handle cyclical dependencies better, but thanks for the suggestion!

guor8lei commented on src/manipulators/no_implicit_any_manipulator.ts:
```diff
@@ -26,9 +37,286 @@ export class NoImplicitAnyManipulator extends Manipulator {
+    // Set of declarations with implicit any type.
+    const modifiedDeclarations = new Set<AcceptedDeclaration>();
+
+    // Iterate through each node, adding declarations to set.
```
Made function, called `getDeclarations()`

guor8lei commented on src/manipulators/no_implicit_any_manipulator.ts:
```diff
@@ -26,9 +37,286 @@ export class NoImplicitAnyManipulator extends Manipulator {
+    >();
+
+    modifiedDeclarations.forEach(declaration => {
+      // For parameter declarations, go though each call expression of the parent function.
```
Broke out into function `addFunctionCallDeclarationDependencies()`

guor8lei commented on src/manipulators/no_implicit_any_manipulator.ts:
```diff
@@ -26,9 +37,286 @@ export class NoImplicitAnyManipulator extends Manipulator {
+        const errorSymbol = errorNode.getSymbol();
+        const declarations = errorSymbol?.getDeclarations();
+        declarations?.forEach(declaration => {
+          if (
```
Changed to use `filter()` instead

guor8lei commented on src/manipulators/manipulator.ts:
```diff
@@ -59,4 +59,45 @@ export abstract class Manipulator {
+   * @return {Statement|undefined} Closest Statement ancestor of modified node or undefined if doesn't exist.
+   */
+  getModifiedStatement(node: Node<ts.Node>): Statement | undefined {
+    return node.getParentWhile((parent, child) => {
```
That's a good point, thanks for catching that! I've added a check to ensure that if the node itself is a Statement, then it is returned.

guor8lei commented on src/manipulators/manipulator.ts:
```diff
@@ -59,4 +59,45 @@ export abstract class Manipulator {
+   * @param {V} val - Value to insert.
+   */
+  addToMapSet<K, V>(map: Map<K, Set<V>>, key: K, val: V): void {
+    if (map.has(key)) {
```
Ohh that's a really good point, I've never really considered that before but it makes a lot of sense. I've gone ahead and changed it to your way because it also looks cleaner too :)).

guor8lei commented on src/manipulators/no_implicit_any_manipulator.ts:
```diff
@@ -26,9 +37,286 @@ export class NoImplicitAnyManipulator extends Manipulator {
+  }
+
+  /**
+   * Adds a new dependency (edge) between two declarations.
```
I added examples of the two main cases where we construct a dependency between two declarations to the comments!

guor8lei commented on test/no_implicit_any_test.ts:
```diff
@@ -0,0 +1,142 @@
+        ['z', new Set(['w'])],
+        ['w', new Set(['y'])],
+      ]),
+      sorted: ['x'],
```
I ended up reimplementing the toposort so that instead of erroring out for cyclical dependencies, it returns an order of declarations that minimizes dependency violations. I also changed the way we're calculating the dependency types so that even in the case of cyclical dependencies, it should be able to add the right types. For the example you gave, the tool would assign type `string` to all 3 variables: y, z, w.

guor8lei commented on src/manipulators/no_implicit_any_manipulator.ts:
```diff
@@ -26,9 +37,286 @@ export class NoImplicitAnyManipulator extends Manipulator {
+      }
+    });
+
+    // If there is a cycle in dependency graph, output to user.
```
Deleted this console log!

guor8lei commented:
A few changes I made while addressing review comments!

- I ended up reimplementing the toposort so that instead of erroring out for cyclical dependencies, it returns an order of declarations that minimizes dependency violations. I also changed the way we're calculating the dependency types so that even in the case of cyclical dependencies, it should be able to add the right types. For example, in the case we have something like: 

```
let y,z,w;

y=z
z=w
w=y
y="string"
```
 the tool would assign type string to all 3 variables: y, z, w.

- Refactored and split the functionality into multiple chunks of functions. I plan on adding unit tests for each of these functions soon!
- Added chalk for color coded error message displays.

jiady commented on src/manipulators/no_implicit_any_manipulator.ts:
```diff
@@ -26,9 +39,391 @@ export class NoImplicitAnyManipulator extends Manipulator {
+    dependencyGraph: Map<AcceptedDeclaration, Set<AcceptedDeclaration>>
+  ) {
+    // Runs code block twice to handle cyclical dependencies between declarations.
+    _.times(2, () => {
```
let's construct a Map<node, set<node>>.

guor8lei replied:
Reimplemented `calculateDeclarationTypes` using the descendants map as we discussed.

guor8lei commented:
Addressed all the comments and added all unit test cases!
